# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

In [8]:
import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv"
df = pd.read_csv(url)

In [9]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     10910 non-null  int64  
 1   Customer                       10910 non-null  object 
 2   State                          10279 non-null  object 
 3   Customer Lifetime Value        10910 non-null  float64
 4   Response                       10279 non-null  object 
 5   Coverage                       10910 non-null  object 
 6   Education                      10910 non-null  object 
 7   Effective To Date              10910 non-null  object 
 8   EmploymentStatus               10910 non-null  object 
 9   Gender                         10910 non-null  object 
 10  Income                         10910 non-null  int64  
 11  Location Code                  10910 non-null  object 
 12  Marital Status                 10910 non-null 

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

In [10]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
df['response'] = df['response'].str.strip().str.lower()


In [ ]:
df = df.drop(columns=['unnamed:_0']) 

In [12]:
new_df = df[
    (df['total_claim_amount'] < 1000) &
    (df['response'] == 'yes')
].copy()

In [13]:
new_df.shape
new_df.head()

customer       state  customer_lifetime_value response  coverage education  \
3   XL78013      Oregon             22332.439460      yes  Extended   College   
8   FM55990  California              5989.773931      yes   Premium   College   
15  CW49887  California              4626.801093      yes     Basic    Master   
19  NJ54277  California              3746.751625      yes  Extended   College   
27  MQ68407      Oregon              4376.363592      yes   Premium  Bachelor   

   effective_to_date employmentstatus gender  income  ...  \
3            1/11/11         Employed      M   49078  ...   
8            1/19/11         Employed      M   66839  ...   
15           1/16/11         Employed      F   79487  ...   
19           2/26/11         Employed      F   41479  ...   
27           2/28/11         Employed      F   63774  ...   

   number_of_open_complaints number_of_policies     policy_type        policy  \
3                        0.0                  2  Corporate Auto  Corporate L3   
8                        0.0                  1   Personal Auto   Personal L1   
15                       0.0                  1    Special Auto    Special L1   
19                       1.0                  1   Personal Auto   Personal L2   
27                       0.0                  1   Personal Auto   Personal L3   

    renew_offer_type  sales_channel  total_claim_amount  vehicle_class  \
3             Offer2         Branch          484.013411  Four-Door Car   
8             Offer2         Branch          739.200000     Sports Car   
15            Offer2         Branch          547.200000            SUV   
19            Offer2    Call Center           19.575683   Two-Door Car   
27            Offer2          Agent           60.036683  Four-Door Car   

   vehicle_size vehicle_type  
3       Medsize            A  
8       Medsize          NaN  
15      Medsize          NaN  
19        Large            A  
27      Medsize          NaN  

[5 rows x 25 columns]

2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

In [15]:
df['response'] = df['response'].str.strip().str.lower()  
df_yes = df[df['response'] == 'yes'].copy()


In [16]:
pivot_metrics = pd.pivot_table(
    df_yes,
    values=['monthly_premium_auto', 'customer_lifetime_value'],
    index='policy_type',
    columns='gender',
    aggfunc='mean'
).round(2)

pivot_metrics

customer_lifetime_value          monthly_premium_auto       
gender                               F        M                    F      M
policy_type                                                                
Corporate Auto                 7712.63  7944.47                94.30  92.19
Personal Auto                  8339.79  7448.38                99.00  91.09
Special Auto                   7691.58  8247.09                92.31  86.34

In [17]:
pivot_claims = pd.pivot_table(
    df_yes,
    values='total_claim_amount',
    index='policy_type',
    columns='gender',
    aggfunc='mean'
).round(2)

pivot_claims


gender               F       M
policy_type                   
Corporate Auto  433.74  408.58
Personal Auto   452.97  457.01
Special Auto    453.28  429.53

Los segmentos Corporate Auto (ambos géneros) y Special Auto (hombres) son los más rentables y de menor riesgo.

Personal Auto presenta mayor riesgo de reclamos, aunque con CLV y primas altas.

La segmentación por género y tipo de póliza permite personalizar campañas y optimizar la rentabilidad del portafolio de clientes.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [18]:
customers_by_state = df['state'].value_counts()
customers_by_state


state
California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: count, dtype: int64

In [19]:
states_over_500 = customers_by_state[customers_by_state > 500]
states_over_500


state
California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: count, dtype: int64

No hay ningún cambio porque todos los  estados tienen más de 500 clientes.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [20]:
clv_stats = df.groupby(['education', 'gender'])['customer_lifetime_value'].agg(['max', 'min', 'median']).round(2)
clv_stats


max      min   median
education            gender                            
Bachelor             F       73225.96  1904.00  5640.51
                     M       67907.27  1898.01  5548.03
College              F       61850.19  1898.68  5623.61
                     M       61134.68  1918.12  6005.85
Doctor               F       44856.11  2395.57  5332.46
                     M       32677.34  2267.60  5577.67
High School or Below F       55277.45  2144.92  6039.55
                     M       83325.38  1940.98  6286.73
Master               F       51016.07  2417.78  5729.86
                     M       50568.26  2272.31  5579.10

Para los clientes con nivel educativo Bachelor, la mediana del CLV es similar entre mujeres (5,640.51) y hombres (5,548.03), aunque los valores máximos son mayores en mujeres, mostrando que algunas generan un valor excepcionalmente alto. En el grupo College, la mediana de los hombres (6,005.85) es ligeramente superior a la de las mujeres (5,623.61), mientras que los máximos son altos para ambos géneros.

En el caso de los clientes con nivel Doctor, la mediana se mantiene similar entre ambos géneros (F: 5,332.46, M: 5,577.67), aunque los valores máximos son menores que en otros grupos, indicando menos variabilidad en los extremos. Por otro lado, los clientes con educación High School or Below muestran medianas superiores (F: 6,039.55, M: 6,286.73) y un valor máximo muy alto en hombres (83,325.38), lo que demuestra que un nivel educativo menor no necesariamente implica un CLV bajo. Finalmente, los clientes con nivel Master presentan valores medianos y máximos equilibrados entre mujeres y hombres, con medianas de 5,729.86 y 5,579.10 respectivamente.

En conclusión, existe gran variabilidad dentro de cada grupo educativo, y no se observa una relación lineal entre mayor educación y CLV más alto. Las medianas son generalmente similares entre géneros, aunque los valores máximos pueden variar significativamente. Esto sugiere que la estrategia de fidelización y upselling debería centrarse en identificar a los clientes de alto CLV dentro de cada grupo, independientemente de su nivel educativo, y considerar otros factores como tipo de póliza, ingresos y canal de venta para una segmentación más efectiva.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [14]:
# your code goes here